rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,329 @@
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Handling out-of-order insertions is not in scope for this RFC, though we make sure the design
+  accounts for the future possibility (see rejected alternatives section)
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
+  will require additional metadata per-SST and a compaction strategy to match, but is otherwise
+  possible with what's outlined in the scope of this design.
+- Outlining compaction strategies that target efficient deletes (such as outlined
+  in [Lethe](https://disc-projects.bu.edu/lethe/))
+- Ensuring deletions are resolved within a time bound (a TTL does not ensure the data is deleted
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+Note that the clock is expected to return monotonically increasing (greater than or equal to
+previous) values. This will be enforced by wrapping the supplied `Clock` implementation that
+tracks the previous values returned.
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    
+    /// returns a timestamp in milliseconds since the unix epoch,
+    /// must return monotonically increasing numbers (this is enforced
+    /// at runtime and will panic if the invariant is broken)
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
+
+    /// the default time-to-live (TTL) for insertions (note that re-inserting a key
+    /// with any value will update the TTL to use the default_ttl)
+    ///
+    /// default: no TTL (insertions will remain until deleted)
+    default_ttl: Option<Duration>
+}
+```
+
+To enforce that the clock remains monotonically increasing across instantiations of the
+database, the manifest will be updated to include the time it was written and this will
+seed the previously returned tick in the clock to the value in the latest manifest:
+
+```fbs
+table ManifestV1 {
```

Comment:
Where would DB level TTL be stored? Or is it a value provided while opening the DB?

Reply:
IIUC it's provided when starting the DB. I asked @agavra to clarify that too (https://github.com/slatedb/slatedb/pull/228#discussion_r1777851299).